### PR TITLE
Add documentation to run Kani on single harnesses

### PR DESCRIPTION
**Pull request by QinyuanWu**
This PR adds to the documentation in the Verification Tools->Kani section of the [challenge book](https://model-checking.github.io/verify-rust-std/tools/kani.html).

### Changes
- Step 1: add clarify on where to put the example code.
- Step 2: add `-Z mem-predicate` to avoid unstable feature errors.
- Step 4: new step to help kani user run specific proofs and identify harness fullnames.

### :exclamation: Issue
I tried to follow step 1 & 2 but kani could not find the harnesses in the example code. At this point there are many proofs in the library so I couldn't find the fullname of the harness in the example code by just running all proofs. I tried to move `example.rs` into `library/core/src/ptr/` and use `--harness ptr::verify::example` or `--harness dummy_proof` but both got `error: no harnesses matched the harness filter`.  Please let me know if you are successful or running into the same issue.

### CODE REVIEW

**Review by danielhumanmod — APPROVED** [review]
LGTM

File: doc/src/tools/kani.md [hunk 1]
```diff
@@ -88,14 +88,14 @@ pub mod verify {
 
 Run the following command in your local terminal:
 
-`kani verify-std -Z unstable-options "path/to/library" --target-dir "/path/to/target" -Z function-contracts -Z stubbing`.
+`kani verify-std -Z unstable-options "path/to/library" --target-dir "/path/to/target" -Z function-contracts -Z stubbing -Z mem-predicates`.
```
Comment (tautschnig) [inline doc/src/tools/kani.md:91]: Yes, thank you, you are right. But then I do wonder whether `-Z stubbing` is even necessary.

**Comment by tautschnig** [comment]
Thank you for your contribution!

> ### ❗ Issue
> 
> I tried to follow step 1 & 2 but kani could not find the harnesses in the example code. At this point there are many proofs in the library so I couldn't find the fullname of the harness in the example code by just running all proofs. I tried to move `example.rs` into `library/core/src/ptr/` and use `--harness ptr::verify::example` or `--harness dummy_proof` but both got `error: no harnesses matched the harness filter`. Please let me know if you are successful or running into the same issue.

Is this still true despite the "Step 4" that you added?

**Comment by QinyuanWu** [comment]
> Thank you for your contribution!
> 
> > ### ❗ Issue
> > I tried to follow step 1 & 2 but kani could not find the harnesses in the example code. At this point there are many proofs in the library so I couldn't find the fullname of the harness in the example code by just running all proofs. I tried to move `example.rs` into `library/core/src/ptr/` and use `--harness ptr::verify::example` or `--harness dummy_proof` but both got `error: no harnesses matched the harness filter`. Please let me know if you are successful or running into the same issue.
> 
> Is this still true despite the "Step 4" that you added?

Yes I get the same error message with this command. `--harness` works for other existing proof harnesses however, such as `--harness ptr::unique::verify` for all proofs in `[unique.rs](https://github.com/model-checking/verify-rust-std/blob/main/library/core/src/ptr/unique.rs)`.

File: doc/src/tools/kani.md (hunk 1 above)
Comment (QinyuanWu) [reply doc/src/tools/kani.md:91]: Removing `-Z stubbing` does not create an issue for the compilation process so I think it's safe to remove it.

**Comment by jaisnan** [comment]
> > Thank you for your contribution!
> > > ### ❗ Issue
> > > I tried to follow step 1 & 2 but kani could not find the harnesses in the example code. At this point there are many proofs in the library so I couldn't find the fullname of the harness in the example code by just running all proofs. I tried to move `example.rs` into `library/core/src/ptr/` and use `--harness ptr::verify::example` or `--harness dummy_proof` but both got `error: no harnesses matched the harness filter`. Please let me know if you are successful or running into the same issue.
> > 
> > 
> > Is this still true despite the "Step 4" that you added?
> 
> @tautschnig Yes I get the same error message with this command. `--harness` works for other existing proof harnesses however, such as `--harness ptr::unique::verify` for all proofs in [unique.rs](https://github.com/model-checking/verify-rust-std/blob/main/library/core/src/ptr/unique.rs).

This might be because --harness and --harness --exact are for individual harnesses and exact harnesses. --harnesses can be used with prefixes, buts it does not verify all harness in a module if you specify it. This is very much an active design choice. If you'd like to change that to recognize all harness in a file or module, you can file a feature request on the kani page :) 

Can you specify which exact harness you want to verify as opposed to stopping at the module, and see if that fixes the error? We can clarify that in the documentation.

**Comment by QinyuanWu** [comment]
@jaisnan I could not get Kani to locate any proofs in `example.rs`. If it works for you, could you please tell me where did you put `example.rs`(which is the sample code in the tutorial) and the complete command you used to locate and verify these proofs?

**Comment by jaisnan** [comment]
> @jaisnan I could not get Kani to locate any proofs in `example.rs`. If it works for you, could you please tell me where did you put `example.rs`(which is the sample code in the tutorial) and the complete command you used to locate and verify these proofs?

I did not put these proofs in a new file no, I think the documentation should make it clear that you should try it out by putting the proofs inside an existing file like `library/std/src/lib.rs`. The starter code is outdated too, since we are now asking users to fork our repo with Kani already initiated in it, and not a fresh copy of the std library. If you don't mind, can I issue a seperate PR that fixes the starting example, and you can simply add the new sections in your PR?

**Review by jaisnan — APPROVED** [review]
Ty! Lgtm now.

File: doc/src/tools/kani.md [hunk 2]
```diff
@@ -97,7 +98,7 @@ The command `kani verify-std` is a sub-command of the `kani`. This specific sub-
 - `"path/to/library"`: This argument specifies the path to the modified Rust Standard Library that was prepared earlier in the script. For example, `./library` or `/home/ubuntu/verify-rust-std/library`
 - `--target-dir "path/to/target"`: This optional argument sets the target directory where Kani will store its output and intermediate files. For example, `/tmp` or `/tmp/verify-std`
 
-Apart from these, you can use your regular `kani-args` such as `-Z function-contracts` and `-Z stubbing` depending on your verification needs.
+Apart from these, you can use your regular `kani-args` such as `-Z function-contracts`, `-Z stubbing` and `-Z mem-predicates` depending on your verification needs. If you run into kani error that says `Use of unstable feature`, add the corresponding feature with `-Z` to the command.
```
Comment (feliperodri) [inline doc/src/tools/kani.md:101]:
```suggestion
Apart from these, you can use your regular `kani-args` such as `-Z function-contracts`, `-Z stubbing` and `-Z mem-predicates` depending on your verification needs. If you run into a Kani error that says `Use of unstable feature`, add the corresponding feature with `-Z` to the command line.
```

File: doc/src/tools/kani.md [hunk 3]
```diff
@@ -143,6 +144,15 @@ Verification Time: 0.01885804s
 Complete - 1 successfully verified harnesses, 0 failures, 1 total.
 ```
 
+Now you can write proof harnesses to verify specific functions in the library. 
+The current convention is to keep proofs in the same module file of the verification target. 
+To run kani for individual proof, use `--harness [harness_function_name]`.
```
Comment (feliperodri) [inline doc/src/tools/kani.md:149]:
```suggestion
To run Kani for an individual proof, use `--harness [harness_function_name]`. 
```

File: doc/src/tools/kani.md [hunk 4]
```diff
@@ -143,6 +144,15 @@ Verification Time: 0.01885804s
 Complete - 1 successfully verified harnesses, 0 failures, 1 total.
 ```
 
+Now you can write proof harnesses to verify specific functions in the library. 
+The current convention is to keep proofs in the same module file of the verification target. 
+To run kani for individual proof, use `--harness [harness_function_name]`. 
+Note that Kani will batch run all proofs in the library folder if you do not supply the `--harness` flag. 
+If kani returns the error `no harnesses matched the harness filter`, you can give the fullname of the harness.
```
Comment (feliperodri) [inline doc/src/tools/kani.md:151]:
```suggestion
If Kani returns the error `no harnesses matched the harness filter`, you can give the full name of the harness. 
```

File: doc/src/tools/kani.md [hunk 5]
```diff
@@ -143,6 +144,15 @@ Verification Time: 0.01885804s
...
 ```
 
+Now you can write proof harnesses to verify specific functions in the library. 
+The current convention is to keep proofs in the same module file of the verification target. 
+To run kani for individual proof, use `--harness [harness_function_name]`. 
+Note that Kani will batch run all proofs in the library folder if you do not supply the `--harness` flag. 
+If kani returns the error `no harnesses matched the harness filter`, you can give the fullname of the harness. 
+For example, to run the proof harness named `check_new` in `library/core/src/ptr/unique.rs`, use 
+`--harness ptr::unique::verify::check_new`. To run all proofs in `unique.rs`, use `--harness ptr::unique::verify`. 
+To find the fullname of a harness, check the kani output and find the line starting with `Checking harness [harness fullname]`.
```
Comment (feliperodri) [inline doc/src/tools/kani.md:154]:
```suggestion
To find the full name of a harness, check the Kani output and find the line starting with `Checking harness [harness full name]`.
```

**Review by feliperodri — APPROVED** [review]

**Review by carolynzech — APPROVED** [review]

**Pull request merged**